{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4206123c",
   "metadata": {},
   "source": [
    "# RootCauseDiscovery (seperate the two datasets and try more patients with ground truth in the appendix as Ines suggested)\n",
    "\n",
    "Here for a known \"root cause gene\", lets run a Lasso regression to filter down the number of features, before running Root Cause Discovery algorithm\n",
    "\n",
    "+ After processing, we obtained 2 datasets\n",
    "    1. \"Interventional data\": there are 32 samples for which we have ground truth data. Each row is a gene and each column is a sample. \n",
    "    2. \"observational data\": these are a few hundred samples for which we did not do any intervention. Each row is a gene and each column is a sample.\n",
    "+ The main idea is that an intervention supposedly caused a disturbance in gene expression, so we can compare the data of intervention to baseline (i.e. observational data), and try to which gene is most differentially expressed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "30dbda32",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "make_plot (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using RootCauseDiscovery\n",
    "using DataFrames\n",
    "using CSV\n",
    "using DelimitedFiles\n",
    "using JLD2\n",
    "using Random\n",
    "using ProgressMeter\n",
    "using LinearAlgebra\n",
    "BLAS.set_num_threads(1)\n",
    "\n",
    "using Plots\n",
    "gr(fmt=:png)\n",
    "\n",
    "# source: Table 1 of https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-022-01019-9\n",
    "# patient_RNA_defects = Dict{String, String}(\n",
    "#     # table S2\n",
    "#     \"R62943\" => \"AE,AS\",\n",
    "#     \"R98254\" => \"AE\",\n",
    "#     \"R86287\" => \"AS\",\n",
    "#     \"R89912\" => \"AE,MAE\",\n",
    "#     \"R19100\" => \"AE\",\n",
    "#     \"R15264\" => \"AE,AS,Var\",\n",
    "#     \"R36605\" => \"AE,AS\",\n",
    "#     \"R61100\" => \"AE,AS\",\n",
    "#     \"R77611\" => \"AE,MAE\",\n",
    "#     \"R16472\" => \"AE,AS,Var\",\n",
    "#     \"R51757\" => \"AE\",\n",
    "#     \"R80346\" => \"AE,Var\",\n",
    "#     \"R20754\" => \"AE\",\n",
    "#     \"R25473\" => \"AE,Var\",\n",
    "#     \"R28774\" => \"AE,AS,Var\",\n",
    "#     \"R96820\" => \"AE,AS\",\n",
    "#     \"R21147\" => \"AE\",\n",
    "#     \"R64921\" => \"AE,AS\",\n",
    "#     \"R52016\" => \"AE,AS,Var\",\n",
    "#     \"R46723\" => \"AE,AS,Var\",\n",
    "#     \"R58859\" => \"AS\",\n",
    "#     \"R80184\" => \"AE,MAE\",\n",
    "#     \"R59185\" => \"AE\",\n",
    "#     \"R63087\" => \"AS\",\n",
    "#     \"R44456\" => \"AE,AS\",\n",
    "#     \"R33391\" => \"AS,Var\",\n",
    "#     \"R66696\" => \"AS\",\n",
    "#     \"R24289\" => \"AE,MAE\",\n",
    "#     \"R98349\" => \"AS\",\n",
    "#     \"R91273\" => \"AE\",\n",
    "#     \"R60537\" => \"AE,Var\",\n",
    "#     \"R70961\" => \"AS\"\n",
    "# )\n",
    "\n",
    "# helper function to submit 1 job to run 1 command\n",
    "function submit(command::String, ncores::Int, total_mem::Number, joblog_dir::String; \n",
    "        jobname=\"submit\", waitfor=Int[], verbose=true)\n",
    "    mem = round(Int, total_mem / ncores) # memory per core\n",
    "    filename = \"$jobname.sh\"\n",
    "    open(filename, \"w\") do io\n",
    "        println(io, \"#!/bin/bash\")\n",
    "        println(io, \"#\")\n",
    "        println(io, \"#SBATCH --job-name=$jobname\")\n",
    "        println(io, \"#\")\n",
    "        println(io, \"#SBATCH --time=48:00:00\")\n",
    "        println(io, \"#SBATCH --cpus-per-task=$ncores\")\n",
    "        println(io, \"#SBATCH --mem-per-cpu=$(mem)G\")\n",
    "        println(io, \"#SBATCH --partition=candes,zihuai,normal,qsu,owners\")\n",
    "        println(io, \"#SBATCH --output=$(joinpath(joblog_dir, \"slurm-%j.out\"))\")\n",
    "        println(io, \"\")\n",
    "        println(io, \"#save job info on joblog:\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID started on:   \\\" `hostname -s`\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID started on:   \\\" `date `\")\n",
    "        println(io, \"\")\n",
    "        println(io, \"# load the job environment:\")\n",
    "        println(io, \"module load julia/1.10\")\n",
    "        println(io, \"export JULIA_DEPOT_PATH=\\\"/home/groups/sabatti/.julia\\\"\")\n",
    "        println(io, \"\")\n",
    "        println(io, \"# run code\")\n",
    "        println(io, \"echo \\\"$command\\\"\")\n",
    "        println(io, \"$command\")\n",
    "        println(io, \"\")\n",
    "        println(io, \"#echo job info on joblog:\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID ended on:   \\\" `hostname -s`\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID ended on:   \\\" `date `\")\n",
    "        println(io, \"#echo \\\" \\\"\")\n",
    "    end\n",
    "    # submit job and capture job ID\n",
    "    io = IOBuffer()\n",
    "    if length(waitfor) != 0\n",
    "        run(pipeline(`sbatch --dependency=afterok:$(join(waitfor, ':')) $filename`; stdout=io))\n",
    "    else\n",
    "        run(pipeline(`sbatch $filename`; stdout=io))\n",
    "    end\n",
    "    msg = String(take!(io))\n",
    "    verbose && println(stdout, msg)\n",
    "    jobid = parse(Int, strip(msg)[21:end])\n",
    "    # clean up and return job ID\n",
    "    close(io)\n",
    "    rm(filename, force=true)\n",
    "    return jobid\n",
    "end\n",
    "\n",
    "# helper function to submit 1 job to run multiple commands\n",
    "function submit(commands::Vector{String}, ncores::Int, \n",
    "        total_mem::Number, joblog_dir::String; \n",
    "        jobname=\"submit\", waitfor=Int[], verbose=true)\n",
    "    mem = round(Int, total_mem / ncores) # memory per core\n",
    "    filename = \"$jobname.sh\"\n",
    "    open(filename, \"w\") do io\n",
    "        println(io, \"#!/bin/bash\")\n",
    "        println(io, \"#\")\n",
    "        println(io, \"#SBATCH --job-name=$jobname\")\n",
    "        println(io, \"#\")\n",
    "        println(io, \"#SBATCH --time=48:00:00\")\n",
    "        println(io, \"#SBATCH --cpus-per-task=$ncores\")\n",
    "        println(io, \"#SBATCH --mem-per-cpu=$(mem)G\")\n",
    "        println(io, \"#SBATCH --partition=candes,zihuai,normal,qsu,owners\")\n",
    "        println(io, \"#SBATCH --output=$(joinpath(joblog_dir, \"slurm-%j.out\"))\")\n",
    "        println(io, \"\")\n",
    "        println(io, \"#save job info on joblog:\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID started on:   \\\" `hostname -s`\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID started on:   \\\" `date `\")\n",
    "        println(io, \"\")\n",
    "        println(io, \"# load the job environment:\")\n",
    "        println(io, \"module load julia/1.10\")\n",
    "        println(io, \"export JULIA_DEPOT_PATH=\\\"/home/groups/sabatti/.julia\\\"\")\n",
    "        println(io, \"\")\n",
    "        for command in commands\n",
    "            println(io, \"echo \\\"$command\\\"\")\n",
    "            println(io, \"$command\")\n",
    "        end\n",
    "        println(io, \"\")\n",
    "        println(io, \"#echo job info on joblog:\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID ended on:   \\\" `hostname -s`\")\n",
    "        println(io, \"echo \\\"Job \\$JOB_ID ended on:   \\\" `date `\")\n",
    "        println(io, \"#echo \\\" \\\"\")\n",
    "    end\n",
    "    # submit job and capture job ID\n",
    "    io = IOBuffer()\n",
    "    if length(waitfor) != 0\n",
    "        run(pipeline(`sbatch --dependency=afterok:$(join(waitfor, ':')) $filename`; stdout=io))\n",
    "    else\n",
    "        run(pipeline(`sbatch $filename`; stdout=io))\n",
    "    end\n",
    "    msg = String(take!(io))\n",
    "    verbose && println(stdout, msg)\n",
    "    jobid = parse(Int, strip(msg)[21:end])\n",
    "    # clean up and return job ID\n",
    "    close(io)\n",
    "    rm(filename, force=true)\n",
    "    return jobid\n",
    "end\n",
    "\n",
    "function make_plot(\n",
    "        Z_score_rank::AbstractVector, \n",
    "        RCD_rank::AbstractVector, \n",
    "        patient_name::AbstractVector;\n",
    "        plot1_ymax = 5500,\n",
    "        plot2_ymin = -2500,\n",
    "        plot2_ymax = 2500,\n",
    "        title1 = \"Rank of root cause (Z-score vs Cholesky-score)\",\n",
    "        title2 = \"Z-score rank - Cholesky-score rank\"\n",
    "    )\n",
    "    # sort by diff in Z scores\n",
    "    Z_score_diff = Z_score_rank - RCD_rank\n",
    "    perm = sortperm(Z_score_diff)\n",
    "\n",
    "    #\n",
    "    # plot 1\n",
    "    #\n",
    "    plt1 = scatter(\n",
    "        1:length(patient_name), \n",
    "        Z_score_rank[perm], \n",
    "        ylabel=\"rank\", \n",
    "        xticks=(1:length(patient_name), patient_name[perm]),\n",
    "        xrotation = 50,\n",
    "        label = \"Z-score rank\",\n",
    "        title = title1,\n",
    "        legend=:top,\n",
    "        dpi=300,\n",
    "        ylim=(0, plot1_ymax)\n",
    "    )\n",
    "    scatter!(plt1, 1:length(patient_name), RCD_rank[perm], label=\"Cholesky-score rank\")\n",
    "\n",
    "    #\n",
    "    # plot 2\n",
    "    #\n",
    "    palettes = theme_palette(:auto)\n",
    "    my_colors = []\n",
    "    for i in eachindex(Z_score_diff)\n",
    "        if isinf(Z_score_diff[i])\n",
    "            push!(my_colors, :black)\n",
    "            Z_score_diff[i] = plot2_ymin\n",
    "        elseif Z_score_diff[i] > 0\n",
    "            push!(my_colors, palettes[1])\n",
    "        elseif Z_score_diff[i] == 0\n",
    "            push!(my_colors, palettes[3])\n",
    "        else\n",
    "            push!(my_colors, palettes[2])\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # make plot\n",
    "    plt2 = scatter(\n",
    "        1:length(patient_name_success), \n",
    "        Z_score_diff[perm],\n",
    "        title=title2, \n",
    "        ylabel=\"Rank difference\",\n",
    "        xticks=(1:length(patient_name), patient_name[perm]),\n",
    "        xrotation = 50,\n",
    "        ylim=(plot2_ymin, plot2_ymax),\n",
    "        label = nothing,\n",
    "        legend=:outerbottom,\n",
    "        color=my_colors[perm],\n",
    "        dpi=300,\n",
    "    )\n",
    "    hline!(plt2, [0], color=:black, linestyle=:dash, label=nothing)\n",
    "\n",
    "    #\n",
    "    # combine plots\n",
    "    #\n",
    "    return plot(plt1, plt2, layout = (2, 1), size=(800, 600))\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18be94e0",
   "metadata": {},
   "source": [
    "## Read processed data\n",
    "First, read in data and log (base 2) transform them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "45114b38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([6.905019519912375 7.257078563799141 … -0.06477114998921524 5.752340009973989; 7.02426909562131 7.246401013060676 … 0.12757476399859816 5.769481834936712; … ; 6.839604518578518 7.29331359490887 … 2.7416948901919214 5.792997014900723; 6.634964406531482 7.155258338398599 … 1.9075765878191422 5.778777598727033], [7.16242663712475 7.523928621811798 … 0.8758080027262863 5.863970440970483; 7.282548459614407 7.484357868453626 … -0.04128210658791 6.128328625903546; … ; 6.791879945272701 7.149247489954807 … 2.2252196140464853 5.952250834124798; 6.954662707932615 7.206785478347909 … 2.43723143625253 5.900881832561235], \u001b[1m69×6 DataFrame\n",
       "\u001b[1m Row │\u001b[1m Patient ID \u001b[1m Genetic diagnosis \u001b[1m gene_id         \u001b[1m patient column index in ⋯\n",
       "     │\u001b[90m String7    \u001b[90m String15          \u001b[90m String15        \u001b[90m Int64                   ⋯\n",
       "─────┼──────────────────────────────────────────────────────────────────────────\n",
       "   1 │ R62943      C19orf70           ENSG00000174917                          ⋯\n",
       "   2 │ R98254      MRPL38             ENSG00000204316\n",
       "   3 │ R86287      DARS2              ENSG00000117593\n",
       "   4 │ R89912      NFU1               ENSG00000169599\n",
       "   5 │ R19100      SLC25A4            ENSG00000151729                          ⋯\n",
       "   6 │ R15264      TIMMDC1            ENSG00000113845\n",
       "   7 │ R36605      TWNK               ENSG00000107815\n",
       "   8 │ R61100      NAXE               ENSG00000163382\n",
       "   9 │ R77611      DLD                ENSG00000091140                          ⋯\n",
       "  10 │ R16472      MRPS25             ENSG00000131368\n",
       "  11 │ R51757      NDUFA10            ENSG00000130414\n",
       "  ⋮  │     ⋮               ⋮                 ⋮                         ⋮       ⋱\n",
       "  60 │ R30525      AP4B1              ENSG00000134262\n",
       "  61 │ R29620      NBAS               ENSG00000151779                          ⋯\n",
       "  62 │ R64055      TALDO1             ENSG00000177156\n",
       "  63 │ R15748      TANGO2             ENSG00000183597\n",
       "  64 │ R66814      VPS33B             ENSG00000184056\n",
       "  65 │ R77365      SELENON            ENSG00000162430                          ⋯\n",
       "  66 │ R21993      SON                ENSG00000159140\n",
       "  67 │ R42505      UQCRFS1            ENSG00000169021\n",
       "  68 │ R64948      MEPCE              ENSG00000146834\n",
       "  69 │ R21470      SFXN4              ENSG00000183605                          ⋯\n",
       "\u001b[36m                                                   3 columns and 48 rows omitted)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function read_data(low_count, threshold, max_cor, concatenate, dir)\n",
    "#     low_count = 10\n",
    "#     threshold = 0.1\n",
    "#     max_cor = 0.999\n",
    "#     concatenate = \"ns\"\n",
    "#     dir = \"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/data\"\n",
    "\n",
    "    # read processed data\n",
    "    isdir(dir) || mkpath(dir)\n",
    "    trainsform_int_file = joinpath(dir, \"transform_int_$concatenate\")\n",
    "    transform_obs_file = joinpath(dir, \"transform_obs_$concatenate\")\n",
    "    ground_truth_file = joinpath(dir, \"ground_truth_$concatenate\")\n",
    "\n",
    "    if !(isfile(trainsform_int_file) && isfile(transform_obs_file) && isfile(ground_truth_file))\n",
    "        # process raw data & save\n",
    "        transform_int, transform_obs, ground_truth = QC_gene_expression_data(;\n",
    "            low_count = low_count, \n",
    "            threshold = threshold, \n",
    "            max_cor = max_cor, \n",
    "            concatenate = concatenate\n",
    "        )\n",
    "        CSV.write(trainsform_int_file, transform_int)\n",
    "        CSV.write(transform_obs_file, transform_obs)\n",
    "        CSV.write(ground_truth_file, ground_truth)\n",
    "    else\n",
    "        transform_int = CSV.read(trainsform_int_file, DataFrame)\n",
    "        transform_obs = CSV.read(transform_obs_file, DataFrame)\n",
    "        ground_truth = CSV.read(ground_truth_file, DataFrame)\n",
    "    end\n",
    "\n",
    "    # numeric matrices\n",
    "    Xobs = transform_obs[:, 2:end] |> Matrix |> transpose\n",
    "    Xint = transform_int[:, 2:end] |> Matrix |> transpose\n",
    "    \n",
    "    return Xobs, Xint, ground_truth\n",
    "end\n",
    "low_count = 10\n",
    "threshold = 0.1\n",
    "max_cor = 0.999\n",
    "concatenate = \"all\"\n",
    "dir = \"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/data\"\n",
    "Xobs, Xint, ground_truth = read_data(low_count, threshold, max_cor, concatenate, dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22e967bd",
   "metadata": {},
   "source": [
    "## Submit main jobs (1 job for each patient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5eec3cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# put this in script /scratch/users/bbchu/RootCauseDiscovery/run_realData_RCD_ines.jl\n",
    "\n",
    "using RootCauseDiscovery\n",
    "using DataFrames\n",
    "using CSV\n",
    "using DelimitedFiles\n",
    "using JLD2\n",
    "using Random\n",
    "using LinearAlgebra\n",
    "BLAS.set_num_threads(1)\n",
    "\n",
    "# inputs to the script\n",
    "patient_id = ARGS[1]\n",
    "y_idx_z_threshold = parse(Float64, ARGS[2])\n",
    "outfile = ARGS[3]\n",
    "low_count = parse(Int, ARGS[4]) # min number of gene count exceeding `threshold` is needed to prevent gene from being filtered out\n",
    "threshold = parse(Float64, ARGS[5]) # for processing raw data\n",
    "seed = parse(Int, ARGS[6])\n",
    "nshuffles = parse(Int, ARGS[7])\n",
    "method = ARGS[8] # \"cv\" or \"nhalf\"\n",
    "max_cor = parse(Float64, ARGS[9])\n",
    "concatenate = ARGS[10] # \"all\", \"ns\", or \"ss\"\n",
    "\n",
    "# for testing\n",
    "# patient_id = \"R62943\"\n",
    "# y_idx_z_threshold = 1.5\n",
    "# outfile = \"fdsa\"\n",
    "# low_count = 10\n",
    "# threshold = 0.1\n",
    "# seed = 1111\n",
    "# nshuffles = 1\n",
    "# method = \"cv\"\n",
    "# max_cor = 0.999\n",
    "# concatenate = \"all\"\n",
    "\n",
    "# process raw data\n",
    "transform_int, transform_obs, _ = \n",
    "    QC_gene_expression_data(low_count = low_count, \n",
    "    threshold = threshold, max_cor = max_cor, \n",
    "    concatenate = concatenate);\n",
    "\n",
    "# numeric matrices\n",
    "Xobs = transform_obs[:, 2:end] |> Matrix |> transpose\n",
    "Xint = transform_int[:, 2:end] |> Matrix |> transpose\n",
    "i = findfirst(x -> x == patient_id, names(transform_int)[2:end])\n",
    "if !isnothing(i)\n",
    "    Xint_sample = Xint[i, :]\n",
    "\n",
    "    # concat Xobs\n",
    "    nint = size(Xint, 1)\n",
    "    Xobs_full = vcat(Xobs, Xint[setdiff(1:nint, i), :])\n",
    "\n",
    "    # run main alg\n",
    "    Random.seed!(seed)\n",
    "    @time root_cause_score = root_cause_discovery_high_dimensional(\n",
    "        Xobs_full, Xint_sample, method, y_idx_z_threshold=y_idx_z_threshold,\n",
    "        nshuffles=nshuffles\n",
    "    );\n",
    "\n",
    "    # save result\n",
    "    writedlm(outfile, root_cause_score)\n",
    "else\n",
    "    println(\"patient $patient_id not found in interventional samples!\")\n",
    "end\n",
    "\n",
    "println(\"Done!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b78ac112",
   "metadata": {},
   "source": [
    "### Submit jobs for each patient\n",
    "\n",
    "20 runs take ~300 sec, so if a patient requires 4000 tries, it'll take ~16.6h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "297b631b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 52100815\n",
      "\n",
      "Submitted batch job 52100817\n",
      "\n",
      "Submitted batch job 52100819\n",
      "\n",
      "Submitted batch job 52100822\n",
      "\n",
      "Submitted batch job 52100824\n",
      "\n",
      "Submitted batch job 52100826\n",
      "\n",
      "Submitted batch job 52100829\n",
      "\n",
      "Submitted batch job 52100831\n",
      "\n",
      "Submitted batch job 52100832\n",
      "\n",
      "Submitted batch job 52100835\n",
      "\n",
      "Submitted batch job 52100837\n",
      "\n",
      "Submitted batch job 52100840\n",
      "\n",
      "Submitted batch job 52100842\n",
      "\n",
      "Submitted batch job 52100845\n",
      "\n",
      "Submitted batch job 52100847\n",
      "\n",
      "Submitted batch job 52100849\n",
      "\n",
      "Submitted batch job 52100850\n",
      "\n",
      "Submitted batch job 52100852\n",
      "\n",
      "Submitted batch job 52100855\n",
      "\n",
      "Submitted batch job 52100858\n",
      "\n",
      "Submitted batch job 52100860\n",
      "\n",
      "Submitted batch job 52100863\n",
      "\n",
      "Submitted batch job 52100865\n",
      "\n",
      "Submitted batch job 52100868\n",
      "\n",
      "Submitted batch job 52100869\n",
      "\n",
      "Submitted batch job 52100870\n",
      "\n",
      "Submitted batch job 52100873\n",
      "\n",
      "Submitted batch job 52100874\n",
      "\n",
      "Submitted batch job 52100875\n",
      "\n",
      "Submitted batch job 52100877\n",
      "\n",
      "Submitted batch job 52100879\n",
      "\n",
      "Submitted batch job 52100881\n",
      "\n",
      "Submitted batch job 52100882\n",
      "\n",
      "Submitted batch job 52100885\n",
      "\n",
      "Submitted batch job 52100886\n",
      "\n",
      "Submitted batch job 52100889\n",
      "\n",
      "Submitted batch job 52100890\n",
      "\n",
      "Submitted batch job 52100892\n",
      "\n",
      "Submitted batch job 52100894\n",
      "\n",
      "Submitted batch job 52100895\n",
      "\n",
      "Submitted batch job 52100897\n",
      "\n",
      "Submitted batch job 52100898\n",
      "\n",
      "Submitted batch job 52100901\n",
      "\n",
      "Submitted batch job 52100902\n",
      "\n",
      "Submitted batch job 52100904\n",
      "\n",
      "Submitted batch job 52100907\n",
      "\n",
      "Submitted batch job 52100909\n",
      "\n",
      "Submitted batch job 52100910\n",
      "\n",
      "Submitted batch job 52100913\n",
      "\n",
      "Submitted batch job 52100914\n",
      "\n",
      "Submitted batch job 52100915\n",
      "\n",
      "Submitted batch job 52100918\n",
      "\n",
      "Submitted batch job 52100919\n",
      "\n",
      "Submitted batch job 52100921\n",
      "\n",
      "Submitted batch job 52100923\n",
      "\n",
      "Submitted batch job 52100925\n",
      "\n",
      "Submitted batch job 52100926\n",
      "\n",
      "Submitted batch job 52100928\n",
      "\n",
      "Submitted batch job 52100930\n",
      "\n",
      "Submitted batch job 52100933\n",
      "\n",
      "Submitted batch job 52100934\n",
      "\n",
      "Submitted batch job 52100936\n",
      "\n",
      "Submitted batch job 52100938\n",
      "\n",
      "Submitted batch job 52100939\n",
      "\n",
      "Submitted batch job 52100941\n",
      "\n",
      "Submitted batch job 52100944\n",
      "\n",
      "Submitted batch job 52100945\n",
      "\n",
      "Submitted batch job 52100947\n",
      "\n",
      "Submitted batch job 52100949\n",
      "\n",
      "Submitted batch job 52100951\n",
      "\n",
      "Submitted batch job 52100952\n",
      "\n",
      "Submitted batch job 52100953\n",
      "\n",
      "Submitted batch job 52100955\n",
      "\n",
      "Submitted batch job 52100960\n",
      "\n",
      "Submitted batch job 52100962\n",
      "\n",
      "Submitted batch job 52100965\n",
      "\n",
      "Submitted batch job 52100966\n",
      "\n",
      "Submitted batch job 52100969\n",
      "\n",
      "Submitted batch job 52100970\n",
      "\n",
      "Submitted batch job 52100972\n",
      "\n",
      "Submitted batch job 52100974\n",
      "\n",
      "Submitted batch job 52100976\n",
      "\n",
      "Submitted batch job 52100977\n",
      "\n",
      "Submitted batch job 52100979\n",
      "\n",
      "Submitted batch job 52100980\n",
      "\n",
      "Submitted batch job 52100981\n",
      "\n",
      "Submitted batch job 52100982\n",
      "\n",
      "Submitted batch job 52100983\n",
      "\n",
      "Submitted batch job 52100984\n",
      "\n",
      "Submitted batch job 52100985\n",
      "\n",
      "Submitted batch job 52100986\n",
      "\n",
      "Submitted batch job 52100987\n",
      "\n",
      "Submitted batch job 52100988\n",
      "\n",
      "Submitted batch job 52100989\n",
      "\n",
      "Submitted batch job 52100990\n",
      "\n",
      "Submitted batch job 52100991\n",
      "\n",
      "Submitted batch job 52100992\n",
      "\n",
      "Submitted batch job 52100993\n",
      "\n",
      "Submitted batch job 52100994\n",
      "\n",
      "Submitted batch job 52100995\n",
      "\n",
      "Submitted batch job 52100996\n",
      "\n",
      "Submitted batch job 52100997\n",
      "\n",
      "Submitted batch job 52100998\n",
      "\n",
      "Submitted batch job 52100999\n",
      "\n",
      "Submitted batch job 52101000\n",
      "\n",
      "Submitted batch job 52101001\n",
      "\n",
      "Submitted batch job 52101002\n",
      "\n",
      "Submitted batch job 52101003\n",
      "\n",
      "Submitted batch job 52101004\n",
      "\n",
      "Submitted batch job 52101006\n",
      "\n",
      "Submitted batch job 52101007\n",
      "\n",
      "Submitted batch job 52101008\n",
      "\n",
      "Submitted batch job 52101009\n",
      "\n",
      "Submitted batch job 52101010\n",
      "\n",
      "Submitted batch job 52101011\n",
      "\n",
      "Submitted batch job 52101012\n",
      "\n",
      "Submitted batch job 52101013\n",
      "\n",
      "Submitted batch job 52101014\n",
      "\n",
      "Submitted batch job 52101015\n",
      "\n",
      "Submitted batch job 52101016\n",
      "\n",
      "Submitted batch job 52101017\n",
      "\n",
      "Submitted batch job 52101018\n",
      "\n",
      "Submitted batch job 52101021\n",
      "\n",
      "Submitted batch job 52101022\n",
      "\n",
      "Submitted batch job 52101023\n",
      "\n",
      "Submitted batch job 52101025\n",
      "\n",
      "Submitted batch job 52101026\n",
      "\n",
      "Submitted batch job 52101027\n",
      "\n",
      "Submitted batch job 52101028\n",
      "\n",
      "Submitted batch job 52101029\n",
      "\n",
      "Submitted batch job 52101031\n",
      "\n",
      "Submitted batch job 52101032\n",
      "\n",
      "Submitted batch job 52101033\n",
      "\n",
      "Submitted batch job 52101034\n",
      "\n",
      "Submitted batch job 52101036\n",
      "\n",
      "Submitted batch job 52101037\n",
      "\n",
      "Submitted batch job 52101038\n",
      "\n",
      "Submitted batch job 52101039\n",
      "\n",
      "Submitted batch job 52101040\n",
      "\n",
      "Submitted batch job 52101041\n",
      "\n",
      "Submitted batch job 52101042\n",
      "\n",
      "Submitted batch job 52101043\n",
      "\n",
      "Submitted batch job 52101044\n",
      "\n",
      "Submitted batch job 52101045\n",
      "\n",
      "Submitted batch job 52101046\n",
      "\n",
      "Submitted batch job 52101047\n",
      "\n",
      "Submitted batch job 52101048\n",
      "\n",
      "Submitted batch job 52101049\n",
      "\n",
      "Submitted batch job 52101050\n",
      "\n",
      "Submitted batch job 52101051\n",
      "\n",
      "Submitted batch job 52101052\n",
      "\n",
      "Submitted batch job 52101053\n",
      "\n",
      "Submitted batch job 52101054\n",
      "\n",
      "Submitted batch job 52101055\n",
      "\n",
      "Submitted batch job 52101056\n",
      "\n",
      "Submitted batch job 52101057\n",
      "\n",
      "Submitted batch job 52101058\n",
      "\n",
      "Submitted batch job 52101060\n",
      "\n",
      "Submitted batch job 52101061\n",
      "\n",
      "Submitted batch job 52101062\n",
      "\n",
      "Submitted batch job 52101063\n",
      "\n",
      "Submitted batch job 52101064\n",
      "\n",
      "Submitted batch job 52101065\n",
      "\n",
      "Submitted batch job 52101066\n",
      "\n",
      "Submitted batch job 52101067\n",
      "\n",
      "Submitted batch job 52101068\n",
      "\n",
      "Submitted batch job 52101069\n",
      "\n",
      "Submitted batch job 52101070\n",
      "\n",
      "Submitted batch job 52101071\n",
      "\n",
      "Submitted batch job 52101072\n",
      "\n",
      "Submitted batch job 52101073\n",
      "\n",
      "Submitted batch job 52101074\n",
      "\n",
      "Submitted batch job 52101075\n",
      "\n",
      "Submitted batch job 52101076\n",
      "\n"
     ]
    }
   ],
   "source": [
    "julia_exe = \"/scratch/users/bbchu/RootCauseDiscovery/run_realData_RCD_ines.jl\"\n",
    "dirs = [\n",
    "    \"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024\"\n",
    "]\n",
    "joblog_dirs = [joinpath(dirs[i], \"joblogs\") for i in eachindex(dirs)]\n",
    "mkpath.(joblog_dirs)\n",
    "seeds = [8262024]\n",
    "low_counts = [10]\n",
    "thresholds = [0.1]\n",
    "nshuffles = 10\n",
    "y_idx_z_threshold = 1.5\n",
    "method = \"cv\"\n",
    "max_cor = 0.999\n",
    "concatenates = [\"all\", \"ns\", \"ss\"]\n",
    "\n",
    "# patient IDs\n",
    "patient_name = [\"R62943\", \"R98254\", \"R89912\", \"R19100\", \"R15264\", \"R36605\", \"R61100\", \n",
    "    \"R77611\", \"R16472\", \"R51757\", \"R80346\", \"R20754\", \"R25473\", \"R28774\", \"R96820\", \n",
    "    \"R21147\", \"R64921\", \"R52016\", \"R46723\", \"R80184\", \"R59185\", \"R44456\", \"R24289\", \n",
    "    \"R91273\", \"R60537\", \"R72253\", \"R75000\", \"R91016\", \"R82353\", \"R34834\", \"R78764\", \n",
    "    \"R30367\", \"R76358\", \"R12128\", \"R45867\", \"R31640\", \"R95723\", \"R55237\", \"R18626\", \n",
    "    \"R34820\", \"R25912\", \"R11258\", \"R64046\", \"R54158\", \"R19907\", \"R27473\", \"R70186\", \n",
    "    \"R26710\", \"R30525\", \"R29620\", \"R64055\", \"R15748\", \"R66814\", \"R77365\", \"R21993\", \n",
    "    \"R42505\", \"R64948\", \"R21470\"]\n",
    "\n",
    "# for testing\n",
    "# nshuffles = 2\n",
    "# y_idx_z_threshold = 20\n",
    "# patient_name = [\"R62943\", \"R98254\"]\n",
    "\n",
    "for concat in concatenates\n",
    "    for (dir, joblog_dir, seed, low_count, threshold) in zip(dirs, joblog_dirs, seeds, low_counts, thresholds)\n",
    "        for patient in patient_name\n",
    "            outdir = joinpath(dir, concat, method)\n",
    "            isdir(outdir) || mkpath(outdir)\n",
    "\n",
    "            outfile = joinpath(outdir, \"$(patient).csv\")\n",
    "            if !isfile(outfile)\n",
    "                cmd = \"julia -t 24 $julia_exe $patient $y_idx_z_threshold $outfile $low_count $threshold $seed $nshuffles $method $max_cor $concat\"\n",
    "                submit(cmd, 24, 64, joblog_dir, jobname=patient)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c7d7987",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_name = [\"R62943\", \"R98254\", \"R89912\", \"R19100\", \"R15264\", \"R36605\", \"R61100\", \n",
    "    \"R77611\", \"R16472\", \"R51757\", \"R80346\", \"R20754\", \"R25473\", \"R28774\", \"R96820\", \n",
    "    \"R21147\", \"R64921\", \"R52016\", \"R46723\", \"R80184\", \"R59185\", \"R44456\", \"R24289\", \n",
    "    \"R91273\", \"R60537\", \"R72253\", \"R75000\", \"R91016\", \"R82353\", \"R34834\", \"R78764\", \n",
    "    \"R30367\", \"R76358\", \"R12128\", \"R45867\", \"R31640\", \"R95723\", \"R55237\", \"R18626\", \n",
    "    \"R34820\", \"R25912\", \"R11258\", \"R64046\", \"R54158\", \"R19907\", \"R27473\", \"R70186\", \n",
    "    \"R26710\", \"R30525\", \"R29620\", \"R64055\", \"R15748\", \"R66814\", \"R77365\", \"R21993\", \n",
    "    \"R42505\", \"R64948\", \"R21470\"]\n",
    "length(patient_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e5ef09",
   "metadata": {},
   "source": [
    "# Plot result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae85bc07",
   "metadata": {},
   "source": [
    "### 8.26.2024 (`nshuffle=10`, new real data analysis based on Ines’ suggestion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "87dbd1a3",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress: 100%|█████████████████████████████████████████| Time: 0:00:02\u001b[39m\n",
      "\u001b[32mProgress: 100%|█████████████████████████████████████████| Time: 0:00:04\u001b[39m\n",
      "\u001b[32mProgress: 100%|█████████████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "function aggregate_result(concatenate::String, dir;\n",
    "    low_count = 10,\n",
    "    threshold = 0.1,\n",
    "    max_cor = 0.999,\n",
    "    )\n",
    "    cv_outdir = joinpath(dir, \"$concatenate/cv\")\n",
    "    data_dir = joinpath(dir, \"data\")\n",
    "    \n",
    "    # import data\n",
    "    Xobs, Xint, ground_truth = read_data(low_count, threshold, max_cor, concatenate, data_dir)\n",
    "    \n",
    "    # only consider patients that is AE\n",
    "    idx = findall(isone, ground_truth[!, \"is AE\"])\n",
    "    ground_truth_filtered = ground_truth[idx, :]\n",
    "    patient_name_filtered = ground_truth_filtered[!, \"Patient ID\"]\n",
    "\n",
    "    # things to return\n",
    "    df = DataFrame(\n",
    "        \"Patient Name\"=>String[],\n",
    "        \"Z-score rank\"=>Float64[],\n",
    "        \"Cholesky-score rank\"=>Float64[],\n",
    "        \"RC Z-score\"=>Float64[],\n",
    "        \"RC Cholesky-score\"=>Float64[],\n",
    "    )\n",
    "    z_all = Vector{Float64}[]\n",
    "\n",
    "    # compare our method against z score method\n",
    "    @showprogress for (i, id) in enumerate(patient_name_filtered)\n",
    "        try\n",
    "            # compute z-score for current sample\n",
    "            Xint_sample = Xint[i, :]\n",
    "            Xobs_full = vcat(Xobs, Xint[setdiff(1:size(Xint, 1), i), :])\n",
    "            z = RootCauseDiscovery.zscore(Xobs_full, Xint_sample)\n",
    "            push!(z_all, z)\n",
    "            \n",
    "            # compute rank of z-score\n",
    "            patient_root_cause_idx = ground_truth_filtered[i, \"root cause row index in genecounts\"]\n",
    "            patient_root_cause_zscore = z[patient_root_cause_idx]\n",
    "            root_cause_zscore_rank = count(x -> x > patient_root_cause_zscore, z)\n",
    "\n",
    "            # cv result\n",
    "            file = joinpath(cv_outdir, \"$id.csv\")\n",
    "            cholesky_score = readdlm(file)\n",
    "            root_cause_cholesky_score = cholesky_score[patient_root_cause_idx]\n",
    "            root_cause_chol_rank_cv = count(x -> x > root_cause_cholesky_score, cholesky_score)\n",
    "\n",
    "            push!(df, [id, root_cause_zscore_rank, root_cause_chol_rank_cv, \n",
    "                           round(patient_root_cause_zscore, digits=3), \n",
    "                            round(root_cause_cholesky_score, digits=3)])\n",
    "        catch e\n",
    "            println(e)\n",
    "            continue\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # @show length(cholesky_score)\n",
    "    df = df[!, [1, 3, 2, 4, 5]]\n",
    "    sort!(df, \"Cholesky-score rank\")\n",
    "    return df, z_all\n",
    "end\n",
    "dir = \"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024\"\n",
    "df_ns, z_ns = aggregate_result(\"ns\", dir)\n",
    "df_ss, z_ss = aggregate_result(\"ss\", dir)\n",
    "df_all, z_all = aggregate_result(\"all\", dir);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "bd8c4eb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_all = 58×6 DataFrame\n",
      " Row │ Patient Name  Cholesky-score rank  Z-score rank  RC Z-score  RC Cholesky-score  origin table\n",
      "     │ String        Float64              Float64       Float64     Float64            String\n",
      "─────┼──────────────────────────────────────────────────────────────────────────────────────────────\n",
      "   1 │ R19100                        0.0           0.0      24.275              8.858  S2\n",
      "   2 │ R61100                        0.0           1.0      33.222             17.558  S2\n",
      "   3 │ R77611                        0.0           9.0       8.15              13.888  S2\n",
      "   4 │ R16472                        0.0           0.0      39.138             12.179  S2\n",
      "   5 │ R28774                        0.0           0.0     133.791             11.93   S2\n",
      "   6 │ R96820                        0.0        1632.0       2.084             10.357  S2\n",
      "   7 │ R64921                        0.0           2.0      17.351             14.071  S2\n",
      "   8 │ R80184                        0.0          10.0      11.534              7.629  S2\n",
      "   9 │ R60537                        0.0           0.0      47.22              18.659  S2\n",
      "  10 │ R82353                        0.0           0.0      48.863             20.647  S3\n",
      "  11 │ R34834                        0.0        1656.0       2.479             10.928  S3\n",
      "  12 │ R30367                        0.0           0.0      31.859              8.792  S3\n",
      "  13 │ R45867                        0.0           0.0     100.224             12.488  S3\n",
      "  14 │ R31640                        0.0           0.0     241.556             47.58   S4\n",
      "  15 │ R55237                        0.0           0.0     186.17              36.46   S4\n",
      "  16 │ R34820                        0.0           4.0      17.536             12.986  S4\n",
      "  17 │ R64046                        0.0          14.0      11.749              5.361  S4\n",
      "  18 │ R19907                        0.0           0.0      99.924             13.349  S4\n",
      "  19 │ R27473                        0.0          37.0       8.851             15.318  S4\n",
      "  20 │ R30525                        0.0           0.0      42.093              9.517  S4\n",
      "  21 │ R64055                        0.0           0.0      92.016             19.481  S4\n",
      "  22 │ R15748                        0.0           0.0      38.237             17.423  S4\n",
      "  23 │ R66814                        0.0           0.0     108.939             15.951  S4\n",
      "  24 │ R77365                        0.0           0.0      39.386             12.71   S4\n",
      "  25 │ R42505                        0.0           0.0      22.415             12.814  S4\n",
      "  26 │ R62943                        1.0         769.0       2.255              6.42   S2\n",
      "  27 │ R11258                        1.0          30.0       8.691              5.917  S4\n",
      "  28 │ R54158                        1.0           0.0     123.273             12.214  S4\n",
      "  29 │ R21470                        1.0          18.0      11.003              5.316  S4\n",
      "  30 │ R21147                        2.0           0.0      36.306              8.282  S2\n",
      "  31 │ R91273                        2.0           1.0      12.237              4.654  S2\n",
      "  32 │ R52016                        3.0           0.0     121.675              7.18   S2\n",
      "  33 │ R91016                        3.0           1.0      50.364              5.224  S3\n",
      "  34 │ R76358                        3.0           5.0      20.283              6.213  S3\n",
      "  35 │ R95723                        3.0         114.0       4.852              6.433  S4\n",
      "  36 │ R21993                        3.0        2001.0       2.193              6.512  S4\n",
      "  37 │ R25473                        4.0          90.0      11.198              5.288  S2\n",
      "  38 │ R64948                        4.0           8.0       7.963              4.508  S4\n",
      "  39 │ R15264                        6.0           0.0      47.713              6.189  S2\n",
      "  40 │ R36605                        6.0         261.0       5.24               5.855  S2\n",
      "  41 │ R20754                        8.0         717.0       7.508              4.938  S2\n",
      "  42 │ R24289                        8.0          20.0       8.143              4.243  S2\n",
      "  43 │ R44456                        9.0          82.0       8.934              4.369  S2\n",
      "  44 │ R89912                       12.0          59.0       9.123              3.742  S2\n",
      "  45 │ R98254                       17.0        2288.0       2.315              3.221  S2\n",
      "  46 │ R72253                       19.0           1.0      28.752              3.918  S3\n",
      "  47 │ R70186                       25.0           1.0      16.23               3.295  S4\n",
      "  48 │ R80346                       28.0        1798.0       3.079              3.984  S2\n",
      "  49 │ R78764                       33.0           3.0      34.76               2.564  S3\n",
      "  50 │ R29620                       33.0         267.0       4.216              2.95   S4\n",
      "  51 │ R51757                       51.0          28.0      13.199              2.563  S2\n",
      "  52 │ R26710                       76.0         575.0       3.189              1.885  S4\n",
      "  53 │ R59185                       87.0        3056.0       4.16               2.851  S2\n",
      "  54 │ R75000                      114.0         584.0       4.714              2.069  S3\n",
      "  55 │ R18626                      995.0         226.0       4.7                0.766  S4\n",
      "  56 │ R12128                     1450.0        1010.0       6.774              0.853  S3\n",
      "  57 │ R46723                     3419.0        1325.0       5.242              0.459  S2\n",
      "  58 │ R25912                     7423.0        7423.0       1.0                0.0    S4\n"
     ]
    }
   ],
   "source": [
    "# add S2/S3/S4 columns\n",
    "data_dir = RootCauseDiscovery.datadir()\n",
    "genecounts = process_data(RootCauseDiscovery.datadir(), \"all\")\n",
    "S2_file = \"13073_2022_1019_MOESM1_ESM_S2.csv\"\n",
    "S3_file = \"13073_2022_1019_MOESM1_ESM_S3.csv\"\n",
    "S4_file = \"13073_2022_1019_MOESM1_ESM_S4.csv\"\n",
    "df2 = RootCauseDiscovery.process_one_root_cause_ground_truth_table(S2_file, genecounts, data_dir)\n",
    "df3 = RootCauseDiscovery.process_one_root_cause_ground_truth_table(S3_file, genecounts, data_dir)\n",
    "df4 = RootCauseDiscovery.process_one_root_cause_ground_truth_table(S4_file, genecounts, data_dir)\n",
    "S2_samples = df2[!, \"Patient ID\"]\n",
    "S3_samples = df3[!, \"Patient ID\"]\n",
    "S4_samples = df4[!, \"Patient ID\"]\n",
    "supp_col = [\"\" for _ in 1:size(df_all, 1)]\n",
    "for (i, id) in enumerate(df_all[!, \"Patient Name\"])\n",
    "    supp_col[i] = id ∈ S2_samples ? \"S2\" : id ∈ S3_samples ? \"S3\" : \"S4\"\n",
    "end\n",
    "df_all[!, \"origin table\"] = supp_col\n",
    "\n",
    "@show df_all;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa01798a",
   "metadata": {},
   "source": [
    "## Save result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "68245fc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/result_ss.csv\""
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# final ranking\n",
    "CSV.write(\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/result_all.csv\", df_all)\n",
    "CSV.write(\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/result_ns.csv\", df_ns)\n",
    "CSV.write(\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/result_ss.csv\", df_ss)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f306b48",
   "metadata": {},
   "source": [
    "Also save Z-scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6071eee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "writedlm(\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/zscores_all.csv\", hcat(z_all...)')\n",
    "writedlm(\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/zscores_ns.csv\", hcat(z_ns...)')\n",
    "writedlm(\"/scratch/users/bbchu/RootCauseDiscovery/8.26.2024/zscores_ss.csv\", hcat(z_ss...)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c2a3c41",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.0",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
